From 3e29e1971bbb4ab63dafeb592b747ef56bad3534 Mon Sep 17 00:00:00 2001
From: Brian Behlendorf <behlendorf1@llnl.gov>
Date: Sat, 29 Aug 2020 20:57:45 -0700
Subject: [PATCH] Linux 5.9 compat: NR_SLAB_RECLAIMABLE
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Commit dcdc12e added compatibility code to treat NR_SLAB_RECLAIMABLE_B
as if it were the same as NR_SLAB_RECLAIMABLE.  However, the new value
is in bytes while the old value was in pages which means they are not
interchangeable.

The only place the reclaimable slab size is used is as a component of
the calculation done by arc_free_memory().  This function returns the
amount of memory the ARC considers to be free or reclaimable at little
cost.  Rather than switch to a new interface to get this value it has
been removed it from the calculation.  It is normally a minor component
compared to the number of inactive or free pages, and removing it
aligns the behavior with the FreeBSD version of arc_free_memory().

Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Reviewed-by: Coleman Kane <ckane@colemankane.org>
Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>
(backport of upstream commit 3e29e1971bbb4ab63dafeb592b747ef56bad3534)
Signed-off-by: Colin Ian King <colin.king@canonical.com>
Closes #10834

---

Index: zfs-linux-0.8.4/config/kernel-global_page_state.m4
===================================================================
--- zfs-linux-0.8.4.orig/config/kernel-global_page_state.m4
+++ zfs-linux-0.8.4/config/kernel-global_page_state.m4
@@ -94,7 +94,6 @@ AC_DEFUN([ZFS_AC_KERNEL_GLOBAL_ZONE_PAGE
 	ZFS_AC_KERNEL_GLOBAL_PAGE_STATE_ENUM_CHECK([NR_FILE_PAGES])
 	ZFS_AC_KERNEL_GLOBAL_PAGE_STATE_ENUM_CHECK([NR_INACTIVE_ANON])
 	ZFS_AC_KERNEL_GLOBAL_PAGE_STATE_ENUM_CHECK([NR_INACTIVE_FILE])
-	ZFS_AC_KERNEL_GLOBAL_PAGE_STATE_ENUM_CHECK([NR_SLAB_RECLAIMABLE])
 
 	AC_MSG_RESULT(yes)
 ])
@@ -126,8 +125,6 @@ AC_DEFUN([ZFS_AC_KERNEL_GLOBAL_PAGE_STAT
 	    [zone_stat_item], [$LINUX/include/linux/mmzone.h])
 	ZFS_AC_KERNEL_ENUM_MEMBER([NR_INACTIVE_FILE],
 	    [zone_stat_item], [$LINUX/include/linux/mmzone.h])
-	ZFS_AC_KERNEL_ENUM_MEMBER([NR_SLAB_RECLAIMABLE],
-	    [zone_stat_item], [$LINUX/include/linux/mmzone.h])
 
 	ZFS_AC_KERNEL_GLOBAL_ZONE_PAGE_STATE_SANITY
 ])
Index: zfs-linux-0.8.4/module/zfs/arc.c
===================================================================
--- zfs-linux-0.8.4.orig/module/zfs/arc.c
+++ zfs-linux-0.8.4/module/zfs/arc.c
@@ -4858,9 +4858,7 @@ arc_free_memory(void)
 	return (ptob(si.freeram - si.freehigh));
 #else
 	return (ptob(nr_free_pages() +
-	    nr_inactive_file_pages() +
-	    nr_inactive_anon_pages() +
-	    nr_slab_reclaimable_pages()));
+	    nr_inactive_file_pages()));
 
 #endif /* CONFIG_HIGHMEM */
 #else
